### PR TITLE
Update hyper to 0.12

### PR DESCRIPTION
**Pull request by remram44**
This updates from [0.9](https://github.com/pingcap/rust-prometheus/pull/17) to 0.12, based on [this example](https://docs.rs/hyper/0.12.21/hyper/server/index.html#example).

### CODE REVIEW

File: examples/example_hyper.rs [hunk 1]
```diff
@@ -45,24 +46,34 @@ lazy_static! {
...
 
             let metric_families = prometheus::gather();
             let mut buffer = vec![];
+            let encoder = TextEncoder::new();
             encoder.encode(&metric_families, &mut buffer).unwrap();
-            res.headers_mut()
-                .set(ContentType(encoder.format_type().parse::<Mime>().unwrap()));
-            res.send(&buffer).unwrap();
+
+            let mut response = Response::new(Body::from(buffer.clone()));
```
Comment (breeswish) [inline examples/example_hyper.rs]: Maybe better to use the builder style? like

```rust
    Response::builder()
        .header(hyper::header::CONTENT_TYPE, encoder.format_type())
        .body(Body::from(buffer.clone()))
        .unwrap()
```

File: examples/example_hyper.rs [hunk 2]
```diff
@@ -45,24 +46,34 @@ lazy_static! {
...
-        .unwrap();
+    };
+
+    println!("listening addr {:?}", addr);
+    let server = Server::bind(&addr.parse().unwrap())
+        .serve(make_service);
+
+    hyper::rt::run(server.map_err(|e| {
+        eprintln!("server error: {}", e);
+}));
```
Comment (breeswish) [inline examples/example_hyper.rs]: Wrong formatting? Run `make dev`.

**Comment by breeswish** [comment]
The rest looks fine! Thanks a lot

**Comment by remram44** [comment]
Thanks for the feedback! I made the requested changes.

**Comment by brson** [comment]
Looks like CI is still complaining about the formatting. @remram44 can you make the tweak to get it passing? Perhaps you need to add the `--alll` flag.

**Comment by remram44** [comment]
This is not related to my merge request, but I'm happy to commit it here.

**Comment by brson** [comment]
Ah sorry @remram44, but thanks for trying to fix it. Now there's a clippy error! I guess master just has linting errors right now.

**Comment by breeswish** [comment]
Sorry for holding this PR for so long time. I will merge it soon.

**Comment by breeswish** [comment]
@overvenus PTAL

**Comment by remram44** [comment]
> You hereby grant to PingCAP and to recipients of software distributed by PingCAP a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable copyright license to reproduce, prepare derivative works of, publicly display, publicly perform, sublicense, and distribute Your Contributions and such derivative works.

Does that mean PingCAP can use the software including third-party contributions without having to follow the terms of the Apache license themselves?

**Comment by breeswish** [comment]
@remram44 Of course not :) But I'm not an expert for this. @Hoverbear Could you help answer that?

**Comment by brson** [comment]
> Does that mean PingCAP can use the software including third-party contributions without having to follow the terms of the Apache license themselves?

I don't know what the intent of the CLA is but that's how it reads to me.

Curiously, the copyright and patent grants in the CLA are almost identical to Apache 2.0:

> 2. Grant of Copyright License. Subject to the terms and conditions of this License, each Contributor hereby grants to You a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable copyright license to reproduce, prepare Derivative Works of, publicly display, publicly perform, sublicense, and distribute the Work and such Derivative Works in Source or Object form.

The CLA though is not the Apache license. In particular the biggest section, the rules for redistribution is not in the CLA. And since the CLA states you retain your copyright, I don't see what the point of the CLA is except to give PingCAP a non-apache license to use the code how it wants.

Somebody correct me.

**Comment by lucab** [comment]
Any chance to rebase this branch and sort out the CLA difficulties soon-ish?

**Comment by brson** [comment]
There's an issue [here](https://github.com/pingcap/tidb/issues/10742) about the CLA and relicensing.

**Comment by breeswish** [comment]
Closing for https://github.com/pingcap/rust-prometheus/pull/249

**Comment by remram44** [comment]
I see. I will refrain from contributing in the future, until #10742 is fixed.

**Pull request closed without merging**